### PR TITLE
Implemented sinh and cosh

### PR DESCRIPTION
**Pull request by lakshayg**
These commits contains the implementation of the sinh and cosh functions for CPU and GPU with gradients. This solves #7531 partially. The supported datatypes are float, double, complex64 and complex128.

Compilation went through successfully on my system (CPU only) and the following test script seems to be running fine on Python 2.7

```python
from __future__ import print_function
import tensorflow as tf
import numpy as np

x = tf.placeholder(shape=[None, 5], dtype=tf.float32)
y = tf.placeholder(shape=[None, 5], dtype=tf.float32)

W = tf.Variable(tf.random_uniform(shape=[5,5], dtype=tf.float32))
w = tf.cosh(W)
error = tf.reduce_mean((tf.matmul(x, w) - y)**2)
train = tf.train.GradientDescentOptimizer(1e-2).minimize(error)

sess = tf.InteractiveSession()
sess.run(tf.global_variables_initializer())
for i in range(100):
    tmp_x = np.random.rand(10, 5)
    tmp_y = np.random.rand(10, 5)

    sess.run(train, feed_dict={x:tmp_x, y:tmp_y})
print(sess.run(W))
```

### CODE REVIEW

File: tensorflow/core/ops/math_grad_test.cc [hunk 1]
```diff
@@ -500,6 +500,26 @@ TEST_F(MathGradTest, Log1p) {
...
+TEST_F(MathGradTest, Cosh) {
+  auto x = test::AsTensor<float>({-3.f, -2.f, -1.f, 1.f, 2.f, 3.f},
+                                 TensorShape({2, 3}));
+  auto g = [](float x) { return std::sinh(x); };
+  auto dx = test::AsTensor<float>(
+      {g(-3.f), g(-2.f), g(-1.f), g(1.f), g(2.f), g(3.f)}, TensorShape({2, 3}));
+  auto ans = SymGrad("Cosh", x);
+  test::ExpectClose(ans, dx);
+}
+
```
Comment (rmlarsen) [inline tensorflow/core/ops/math_grad_test.cc]: Add tests for complex.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
Looks really good!

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

File: tensorflow/cc/gradients/math_grad_test.cc [hunk 2]
```diff
@@ -337,6 +345,50 @@ TEST_F(CWiseUnaryGradTest, Log1p_Complex) {
...
+
+TEST_F(CWiseUnaryGradTest, Sinh_Complex) {
+  auto x_fn = [this](const int i) {
+    return CRV({{1, 0}, {0, 1}, {2, -1}, {1, 2}, {3, 4}});
+  };
+  auto dy_fn = [this](const complex64& x) {
+    return x + CRV({{-2, 2}, {-3, 3}, {1, -4}});
+  };
+  auto dx_fn = [this](const complex64& x, const complex64& dy) {
+    return dy * conjugate(std::cos(x));
```
Comment (rmlarsen) [inline tensorflow/cc/gradients/math_grad_test.cc]: Ah, it looks like you need to replace cos with cosh here to fix the broken test.

**Comment by rmlarsen** [comment]
@lakshayg It looks like there's a bug in the C++ complex gradient for sinh: [ RUN      ] CWiseUnaryGradTest.Sinh_Complex
2017-06-06 20:16:34.833848: E tensorflow/core/framework/tensor_testutil.cc:34] x = Tensor<type: complex64 shape: [2,3,2] values: ? ? ?...>
2017-06-06 20:16:34.833874: E tensorflow/core/framework/tensor_testutil.cc:35] y = Tensor<type: complex64 shape: [2,3,2] values: ? ? ?...>
2017-06-06 20:16:34.833936: E tensorflow/core/framework/tensor_testutil.cc:36] atol = 1e-06 rtol = 1e-06 tol = 9.1994e-06
tensorflow/core/framework/tensor_testutil.cc:38: Failure
Value of: false
  Actual: false
Expected: true
0-th element is not close (-8.1365184783935547,1.0135481357574463) vs. (2.779362678527832,-0.21568894386291504)
2017-06-06 20:16:34.834010: E tensorflow/core/framework/tensor_testutil.cc:34] x = Tensor<type: complex64 shape: [2,3,2] values: ? ? ?...>
2017-06-06 20:16:34.834015: E tensorflow/core/framework/tensor_testutil.cc:35] y = Tensor<type: complex64 shape: [2,3,2] values: ? ? ?...>
2017-06-06 20:16:34.834021: E tensorflow/core/framework/tensor_testutil.cc:36] atol = 1e-06 rtol = 1e-06 tol = 4.45043e-06
tensorflow/core/framework/tensor_testutil.cc:38: Failure
Value of: false
  Actual: false
Expected: true
1-th element is not close (-1.5430805683135986,3.0861611366271973) vs. (-0.54030227661132812,1.0806045532226562)

File: tensorflow/core/ops/math_grad_test.cc (hunk 1 above)
Comment (lakshayg) [inline tensorflow/core/ops/math_grad_test.cc]: @rmlarsen I am not sure how to implement the tests for complex here as no other functions like sin, cos, tan, tanh do not implement such test. Also, I implemeted the test for complex in ` tensorflow/cc/gradients/math_grad_test.cc`. Did you mean this file?

**Comment by lakshayg** [comment]
@rmlarsen I am also working on implementing the inverse hyperbolic functions and having some issues with understanding the code. It would be great if you could help me by answering the question at https://stackoverflow.com/questions/44389994/making-sense-of-gradforunarycwise-in-tensorflow

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

File: tensorflow/core/ops/math_grad_test.cc (hunk 1 above)
Comment (rmlarsen) [inline tensorflow/core/ops/math_grad_test.cc]: I think it's OK. You have test coverage in several other places.

**Comment by lakshayg** [comment]
I thinks this is somewhat strange. Of the 4 tests which failed, 3 ran successfully (Windows Cmake Tests, Linux GPU, Linux CPU Tests) before the last revision. Since the last revision involved correcting one of the test functions (changing `std::cos(x)` to `std::cosh(x)`), it surprises me that tests in unrelated files are failing.

I ran the following commands on my machine and the tests were successful

1. bazel test //tensorflow/python/kernel_tests:check_ops_test 
2. bazel test //tensorflow/python/kernel_tests:embedding_ops_test
3. bazel test //tensorflow/contrib/seq2seq:beam_search_decoder_test
4. bazel test //tensorflow/python/kernel_tests:sparse_ops_test
5. bazel test //tensorflow/contrib/seq2seq:basic_decoder_test
6. bazel test //tensorflow/contrib/layers:embedding_ops_test

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@lakshayg I think the test failures are unrelated to your change.

**Review by rmlarsen — APPROVED** [review]
